### PR TITLE
DRILL-7984: Support clickhouse by JDBC plugin

### PR DESCRIPTION
**Pull request by Leon-WTF**
# [DRILL-7984](https://issues.apache.org/jira/browse/DRILL-7984): Support clickhouse by JDBC plugin

## Description
1. add clickhouse dialect in drill calcite
2. add clickhouse compatibility in jdbc plugin

## Documentation
TODO update https://drill.apache.org/docs/rdbms-storage-plugin/

## Testing
See TestJdbcPluginWithClickhouse

### CODE REVIEW

File: contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/JdbcCatalogSchema.java [hunk 1]
```diff
@@ -53,11 +54,12 @@
     try (Connection con = source.getConnection();
          ResultSet set = con.getMetaData().getCatalogs()) {
       connectionSchemaName = con.getSchema();
-      while (set.next()) {
-        final String catalogName = set.getString(1);
-        CapitalizingJdbcSchema schema = new CapitalizingJdbcSchema(
-            getSchemaPath(), catalogName, source, dialect, convention, catalogName, null, caseSensitive);
-        schemaMap.put(schema.getName(), schema);
+      if (!ClickhouseConstant.PRODUCT_NAME.equals(con.getMetaData().getDatabaseProductName())) {
```
Comment (paul-rogers) [inline contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/JdbcCatalogSchema.java:57]: Let's think about this approach a bit. There is no comment here (there should be), so I'll infer that Clickhouse somehow does something special. Cool. So I'm working on Postgres, or DB X and I want to do something similar. Do we end up with a bit set of if-statements? If I work on DB X, how do I ensure I don't break Clickhouse, since I can't run that code?

Can we think about a way of abstracting this kind of thing into a DB-specific helper class? The default helper does "stock" JDBC. Clickhouse (and my DB X) code go into DB-specific helpers. Now, to get DB X to work, I don't have to worry about breaking Clickhouse.

See if you can work out a "mini-plugin" for such code. Happy to provide suggestions if you like.

File: contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/JdbcCatalogSchema.java [hunk 2]
```diff
@@ -108,7 +113,7 @@ private boolean addSchemas(DataSource source, SqlDialect dialect, DrillJdbcConve
 
         String parentKey = StringUtils.lowerCase(catalogName);
         CapitalizingJdbcSchema parentSchema = schemaMap.get(parentKey);
-        if (parentSchema == null) {
+        if (parentSchema == null || isFirstLevel) {
```
Comment (paul-rogers) [inline contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/JdbcCatalogSchema.java:116]: Nit: comment to explain why the first level matters.

File: contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/JdbcPrel.java [hunk 3]
```diff
@@ -39,6 +39,9 @@
 import org.apache.drill.exec.planner.physical.visitor.PrelVisitor;
 import org.apache.drill.exec.record.BatchSchema.SelectionVectorMode;
 import org.apache.drill.exec.store.SubsetRemover;
+import org.apache.drill.exec.store.jdbc.clickhouse.ClickhouseConstant;
+import org.apache.drill.exec.store.jdbc.clickhouse.ClickhouseDialect;
+import org.apache.drill.exec.store.jdbc.clickhouse.ClickhouseJdbcImplementor;
```
Comment (paul-rogers) [inline contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/JdbcPrel.java:44]: See comment above. The generic `JdbcPrel` is not the place to add vendor-specific code. However, this class can call a vendor-specific helper.

File: contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/clickhouse/ClickhouseConstant.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.drill.exec.store.jdbc.clickhouse;
+
+/**
+ * @author feiteng.wtf
+ * @date 2021-08-08
+ */
```
Comment (paul-rogers) [inline contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/clickhouse/ClickhouseConstant.java:23]: Generally Drill does not include the author and date. What we do include is an explanation of what the class is about.

File: contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/clickhouse/ClickhouseDialect.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.SqlUserDefinedTypeNameSpec;
+import org.apache.calcite.sql.SqlWriter;
+import org.apache.calcite.sql.parser.SqlParserPos;
+import org.apache.calcite.sql.type.BasicSqlType;
+
+/**
+ * @author feiteng.wtf
+ * @date 2021-08-08
+ */
```
Comment (paul-rogers) [inline contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/clickhouse/ClickhouseDialect.java:33]: See comment above.

File: contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/clickhouse/ClickhouseDialect.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+public class ClickhouseDialect extends SqlDialect {
```
Comment (paul-rogers) [inline contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/clickhouse/ClickhouseDialect.java:34]: This is a great example of a vendor-specific implementation. Can we follow the same pattern elsewhere?

File: contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/clickhouse/ClickhouseJdbcImplementor.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+import org.apache.calcite.adapter.jdbc.JdbcTableScan;
+import org.apache.calcite.sql.SqlDialect;
+import org.apache.calcite.sql.SqlIdentifier;
+
+import java.util.Iterator;
+
+/**
+ * @author feiteng.wtf
+ * @date 2021-07-26
+ */
```
Comment (paul-rogers) [inline contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/clickhouse/ClickhouseJdbcImplementor.java:33]: Ditto.

File: contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/writers/JdbcBigintWriter.java [hunk 8]
```diff
@@ -31,9 +31,9 @@ public JdbcBigintWriter(String colName, RowSetLoader rowWriter, int columnIndex)
 
   @Override
   public void load(ResultSet results) throws SQLException {
-    boolean b = results.wasNull();
-    if (! results.wasNull()) {
-      long value = results.getLong(columnIndex);
+    // clickhouse requires getting the column before checking nullability
+    long value = results.getLong(columnIndex);
+    if (!results.wasNull()) {
```
Comment (paul-rogers) [inline contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/writers/JdbcBigintWriter.java:36]: Good catch! As it turns out, the original code is wrong for all DBs, not just Clickhouse. See [this article](https://schneide.blog/2021/02/22/jdbcs-wasnull-method-pitfall/).

Suggestion, change the comment to something like "JDBC reports nullability only after getting the column value." Here and below.

**Review by paul-rogers — CHANGES_REQUESTED** [review]
Thanks much for the contribution! The logic looks right. My comments are mostly about the "Drill way" of doing things. Each project has its preferences, and I've pointed out Drill's.

File: contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/clickhouse/ClickhouseDialect.java (hunk 6 above)
Comment (cgivre) [reply contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/clickhouse/ClickhouseDialect.java:34]: I'm not sure if this is relevant here or not, but Calcite actually has a Clickhouse dialect.[1]. Does that "answer the mail" for this?

[1]: https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/sql/dialect/ClickHouseSqlDialect.java

**Review by cgivre — COMMENTED** [review]
Could we add some additional unit tests to this PR?

File: contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/clickhouse/ClickhouseDialect.java (hunk 6 above)
Comment (Leon-WTF) [reply contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/clickhouse/ClickhouseDialect.java:34]: thanks, but as it is not included in the current version of calcite, so I will need to do a cherry pick on DrillCalcite1.21.0 of https://github.com/vvysotskyi/drill-calcite.git and update to 1.21.0-drill-r4, right?

Comment (vdiravka) [reply contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/clickhouse/ClickhouseDialect.java:34]: To update Calcite - yes. Please do it via PR
Here is the process how to do that: https://github.com/apache/drill/blob/master/docs/dev/Calcite.md#process-of-updating-calcite-version
You can ask me or @vvysotskyi to follow you in this

**Comment by Leon-WTF** [comment]
> Thanks much for the contribution! The logic looks right. My comments are mostly about the "Drill way" of doing things. Each project has its preferences, and I've pointed out Drill's.

Sure, your comments help me to improve a lot.

File: contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/JdbcCatalogSchema.java (hunk 1 above)
Comment (Leon-WTF) [reply contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/JdbcCatalogSchema.java]: I have added JdbcSchemaFactory and JdbcSqlGenerator, please have a look.

File: contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/JdbcCatalogSchema.java (hunk 2 above)
Comment (Leon-WTF) [reply contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/JdbcCatalogSchema.java]: See above

File: contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/JdbcPrel.java (hunk 3 above)
Comment (Leon-WTF) [reply contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/JdbcPrel.java]: See above

File: contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/clickhouse/ClickhouseDialect.java (hunk 6 above)
Comment (Leon-WTF) [reply contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/clickhouse/ClickhouseDialect.java]: Please have a look at https://github.com/vvysotskyi/drill-calcite/pull/4

File: contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/JdbcSchemaFactory.java [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+public class JdbcSchemaFactory {
+  private final JdbcStoragePlugin plugin;
+
+  public JdbcSchemaFactory (JdbcStoragePlugin plugin) {
+    this.plugin = plugin;
+  }
+
+  public void registerSchemas(SchemaConfig config, SchemaPlus parent) {
+    if (plugin.getConfig().isClickhouse()) {
```
Comment (paul-rogers) [inline contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/JdbcSchemaFactory.java:32]: Here we've added an `isClickhouse()` method to the base JDBC plugin. We might add five more "isDruid()", "isAthena()", "isNextBigThing()", ... Then we add logic in various places that use those queries.

In doing so, we couple all these dialects and make the code harder to maintain.

Suggestion: add a `getDialect()` method to the config. It returns a `JdbcDialect` subclass. `getDialect()` does what is needed to identify Clickhouse (or whatever) and return the specific dialect. Else, it returns a `GenericDialect` instance.

One of the methods on `JdbcDialect` is `registerSchemas()`. The "then" part of this code goes in the Clickhouse version, the "else" part of this code goes into the `GenericDialect` version.

Voila! No more tight coupling of dialects.

Of course, if there is a better way to solve the problem, feel free to use that instead; the above is a quick & dirty example.

File: contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/JdbcSqlGenerator.java [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+public class JdbcSqlGenerator {
+  /**
+   * Generate sql for different jdbc databases
+   */
+  public static String generateSql(JdbcStoragePlugin plugin,
+                                   RelOptCluster cluster, RelNode input) {
+    final SqlDialect dialect = plugin.getDialect();
+    final JdbcImplementor jdbcImplementor;
+    if (plugin.getConfig().isClickhouse()) {
```
Comment (paul-rogers) [inline contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/JdbcSqlGenerator.java:36]: Here, we'd add a `generateSQL` method to our `JdbcDialect` interface, moving this code to subclasses as described above.

File: contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/JdbcSchemaFactory.java (hunk 9 above)
Comment (cgivre) [reply contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/JdbcSchemaFactory.java:32]: @paul-rogers 
Alternatively, the dialect is actually stored in the Plugin file, not the config file.   More good news is that there already is a method in the Plugin file:

https://github.com/apache/drill/blob/39b565f112122734c080324fdcbef518ced16507/contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/JdbcStoragePlugin.java#L81-L83

If I'm tracking the logic correctly, you could use this logic to get the dialect without having to modify the config at all.

Comment (Leon-WTF) [reply contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/JdbcSchemaFactory.java]: @cgivre I think @paul-rogers means a new "dialect" interface to be implemented to handle the differrent behaviors for databases connected using JdbcStoragePlugin, as registerSchemas and generateSql.

File: contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/clickhouse/ClickhouseDialect.java (hunk 6 above)
Comment (vdiravka) [reply contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/clickhouse/ClickhouseDialect.java]: Your Calcite PR is merged, git tag is created and Maven artifacts are available in Jitpack:
https://jitpack.io/#vvysotskyi/drill-calcite

File: logical/src/main/java/org/apache/drill/common/logical/StoragePluginConfig.java [hunk 11]
```diff
@@ -19,11 +19,9 @@
 
 
 import com.fasterxml.jackson.annotation.JsonIgnore;
-import com.fasterxml.jackson.annotation.JsonInclude;
 import com.fasterxml.jackson.annotation.JsonTypeInfo;
 
 @JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
-@JsonInclude(JsonInclude.Include.NON_DEFAULT)
```
Comment (vvysotskyi) [inline logical/src/main/java/org/apache/drill/common/logical/StoragePluginConfig.java:26]: Please revert these changes. It helps to preserve backward compatibility for the case when the new option was added, the user started a newer version of Drill, but after that decided to use previous one.

Comment (Leon-WTF) [reply logical/src/main/java/org/apache/drill/common/logical/StoragePluginConfig.java]: Sure, this is for fixing one ut case, I have tried another way.

**Comment by Leon-WTF** [comment]
> Could we add some additional unit tests to this PR?

Added in TestJdbcPluginWithClickhouse similar to TestJdbcPluginWithMySQLIT

**Review by paul-rogers — APPROVED** [review]
Thanks for creating the JDBC dialect mechanism: it nicely solves the problems we discussed earlier. I reviewed the non-Clickhouse bits of code; they look fine, so LGTM.
+1

**Comment by vvysotskyi** [comment]
@Leon-WTF, looks like this PR breaks Travis CI, could you please fix it?
```
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 144.758 s <<< FAILURE! - in org.apache.drill.exec.store.jdbc.TestJdbcPluginWithClickhouse
3883[ERROR] org.apache.drill.exec.store.jdbc.TestJdbcPluginWithClickhouse  Time elapsed: 144.758 s  <<< ERROR!
3884org.testcontainers.containers.ContainerLaunchException: Container startup failed
3885	at org.apache.drill.exec.store.jdbc.TestJdbcPluginWithClickhouse.initClickhouse(TestJdbcPluginWithClickhouse.java:58)
3886Caused by: org.rnorth.ducttape.RetryCountExceededException: Retry limit hit with exception
3887	at org.apache.drill.exec.store.jdbc.TestJdbcPluginWithClickhouse.initClickhouse(TestJdbcPluginWithClickhouse.java:58)
3888Caused by: org.testcontainers.containers.ContainerLaunchException: Could not create/start container
3889	at org.apache.drill.exec.store.jdbc.TestJdbcPluginWithClickhouse.initClickhouse(TestJdbcPluginWithClickhouse.java:58)
3890Caused by: org.testcontainers.ext.ScriptUtils$UncategorizedScriptException: 
3891Failed to execute database script from resource [create table person (
3892  person_id       Int32,
3893
3894  first_name      Nullable(String),
3895  last_name       Nullable(String),
3896  address         Nullable(String),
3897  city            Nullable(String),
3898  state           Nullable(String),
3899  zip             Nullable(Int32),
3900
3901  json            Nullable(String),
3902
3903  bigint_field    Nullable(Int64),
3904  smallint_field  Nullable(Int16),
3905  decimal_field   Nullable(DECIMAL(15, 2)),
3906  boolean_field   Nullable(UInt8),
3907  double_field    Nullable(Float64),
3908  float_field     Nullable(Float32),
3909
3910  date_field      Nullable(Date),
3911  datetime_field  Nullable(Datetime),
3912  enum_field      Enum('XXX'=1, 'YYY'=2, 'ZZZ'=3)
3913) ENGINE = MergeTree() order by person_id;
3914
3915insert into person (person_id, first_name, last_name, address, city, state, zip, json,
3916                    bigint_field, smallint_field, decimal_field, boolean_field, double_field,
3917                    float_field, date_field, datetime_field, enum_field)
3918values (1, 'first_name_1', 'last_name_1', '1401 John F Kennedy Blvd', 'Philadelphia', 'PA',
3919        19107, '{ a : 5, b : 6 }', 123456789, 1, 123.321, 0, 1.0, 1.1, '2012-02-29',
3920        '2012-02-29 13:00:01', 'XXX');
3921
3922insert into person (person_id, first_name, last_name, address, city, state, zip, json,
3923                    bigint_field, smallint_field, boolean_field, double_field,
3924                    float_field, date_field, datetime_field, enum_field)
3925values (2, 'first_name_2', 'last_name_2', 'One Ferry Building', 'San Francisco', 'CA', 94111,
3926        '{ z : [ 1, 2, 3 ] }', 45456767, 3, 1, 3.0, 3.1, '2011-10-30',
3927        '2011-10-30 11:34:21', 'YYY');
3928
3929insert into person (person_id, first_name, last_name, address, city, state, zip, json,
3930                    bigint_field, smallint_field, boolean_field, double_field,
3931                    float_field, date_field, datetime_field, enum_field)
3932values (3, 'first_name_3', 'last_name_3', '176 Bowery', 'New York', 'NY', 10012, '{ [ a, b, c ] }',
3933        123090, -3, 0, 5.0, 5.1, '2015-06-01', '2015-09-22 15:46:10', 'ZZZ');
3934
3935insert into person (person_id) values (4);
3936
3937create view person_view as select * from person;
3938]
3939	at org.apache.drill.exec.store.jdbc.TestJdbcPluginWithClickhouse.initClickhouse(TestJdbcPluginWithClickhouse.java:58)
3940Caused by: org.testcontainers.exception.ConnectionCreationException: Could not obtain JDBC connection
3941	at org.apache.drill.exec.store.jdbc.TestJdbcPluginWithClickhouse.initClickhouse(TestJdbcPluginWithClickhouse.java:58)
3942Caused by: java.sql.SQLException: Could not create new connection
3943	at org.apache.drill.exec.store.jdbc.TestJdbcPluginWithClickhouse.initClickhouse(TestJdbcPluginWithClickhouse.java:58)
```

**Comment by Leon-WTF** [comment]
> 3942Caused by: java.sql.SQLException: Could not create new connection

Sure, do you know what's the difference betwen Travis CI and GitHub CI? The GitHub CI seems all passed.

**Comment by Leon-WTF** [comment]
> Errors: 1

[ERROR] Crashed tests:
[ERROR] org.apache.drill.exec.fn.impl.TestAggregateFunctions
[ERROR] org.apache.maven.surefire.booter.SurefireBooterForkException: ExecutionException The forked VM terminated without properly saying goodbye. VM crash or System.exit called?
[ERROR] Command was /bin/sh -c cd /home/travis/build/apache/drill/exec/java-exec && /usr/lib/jvm/java-8-openjdk-arm64/jre/bin/java -Xms512m -Xmx1300m -Ddrill.exec.http.enabled=false -Ddrill.exec.memory.enable_unsafe_bounds_check=true -Ddrill.exec.sys.store.provider.local.write=false -Dorg.apache.drill.exec.server.Drillbit.system_options=org.apache.drill.exec.compile.ClassTransformer.scalar_replacement=on -Ddrill.catastrophic_to_standard_out=true -XX:MaxDirectMemorySize=3000M -Djava.net.preferIPv4Stack=true -Djava.awt.headless=true -XX:+CMSClassUnloadingEnabled -ea -Djdk.attach.allowAttachSelf=true -javaagent:/home/travis/.m2/repository/org/jmockit/jmockit/1.47/jmockit-1.47.jar org.apache.maven.surefire.booter.ForkedBooter /home/travis/build/apache/drill/exec/java-exec/target/surefire 2021-09-04T10-46-59_724-jvmRun1 surefire6858063389372217786tmp surefire_132509789057170760425tmp
[ERROR] Error occurred in starting fork, check output in log
[ERROR] Process Exit Code: 255
[ERROR] Crashed tests:
[ERROR] org.apache.drill.exec.fn.impl.TestAggregateFunctions
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.awaitResultsDone(ForkStarter.java:532)
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.runSuitesForkOnceMultiple(ForkStarter.java:405)
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.run(ForkStarter.java:321)
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.run(ForkStarter.java:266)
[ERROR] 	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeProvider(AbstractSurefireMojo.java:1314)
[ERROR] 	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeAfterPreconditionsChecked(AbstractSurefireMojo.java:1159)
[ERROR] 	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.execute(AbstractSurefireMojo.java:932)
[ERROR] 	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:137)
[ERROR] 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:210)
[ERROR] 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:156)
[ERROR] 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:148)
[ERROR] 	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:117)
[ERROR] 	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:81)
[ERROR] 	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:56)
[ERROR] 	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
[ERROR] 	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:305)
[ERROR] 	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:192)
[ERROR] 	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:105)
[ERROR] 	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:957)
[ERROR] 	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:289)
[ERROR] 	at org.apache.maven.cli.MavenCli.main(MavenCli.java:193)
[ERROR] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[ERROR] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR] 	at java.lang.reflect.Method.invoke(Method.java:498)
[ERROR] 	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:282)
[ERROR] 	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:225)
[ERROR] 	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:406)
[ERROR] 	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:347)
[ERROR] Caused by: org.apache.maven.surefire.booter.SurefireBooterForkException: The forked VM terminated without properly saying goodbye. VM crash or System.exit called?


@vvysotskyi I can run TestJdbcPluginWithClickhouse successfully locally, and the error changed after re-run by merge. Could you help to re-run the Travis CI again? I don't have the permission to do that.

**Comment by vvysotskyi** [comment]
@Leon-WTF, the main difference is that Travis build uses an` arm64-graviton2` arch.
Regarding the exception I shared, please see the build for master branch: https://app.travis-ci.com/github/apache/drill/jobs/536454728

**Comment by Leon-WTF** [comment]
> @Leon-WTF, the main difference is that Travis build uses an` arm64-graviton2` arch.
> Regarding the exception I shared, please see the build for master branch: https://app.travis-ci.com/github/apache/drill/jobs/536454728

Seems that the clickhouse 

> @Leon-WTF, the main difference is that Travis build uses an` arm64-graviton2` arch.
> Regarding the exception I shared, please see the build for master branch: https://app.travis-ci.com/github/apache/drill/jobs/536454728

@vvysotskyi Fixed in https://github.com/apache/drill/pull/2315, sorry for the inconvenience~

**Comment by vvysotskyi** [comment]
@Leon-WTF, thanks for fixing it!
